File: app/src/main/java/com/teixeira0x/subtypo/preferences/UIModePreference.java
package com.teixeira0x.subtypo.preferences;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.AttributeSet;
import androidx.preference.Preference;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.teixeira0x.subtypo.App;
import com.teixeira0x.subtypo.R;

public class UIModePreference extends Preference {

  public UIModePreference(Context context) {
    super(context);
  }

  public UIModePreference(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public UIModePreference(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  public UIModePreference(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
    super(context, attrs, defStyleAttr, defStyleRes);
  }

  @Override
  protected void onClick() {
    super.onClick();

    MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(getContext());
    builder.setTitle(getTitle());

    String[] items =
        new String[] {
          getString(R.string.pref_aparence_ui_mode_value_follow),
          getString(R.string.pref_aparence_ui_mode_value_light),
          getString(R.string.pref_aparence_ui_mode_value_dark),
        };

    SharedPreferences.Editor prefEdit = getSharedPreferences().edit();
    int selectedItemIndex = getSharedPreferences().getInt(getKey(), 0);

    builder.setSingleChoiceItems(items, selectedItemIndex, (d, w) -> prefEdit.putInt(getKey(), w));
    builder.setNegativeButton(R.string.cancel, null);
    builder.setPositiveButton(
        R.string.save,
        (d, w) -> {
          prefEdit.apply();
          App.getInstance().updateUIMode();
        });
    builder.show();
  }

  private String getString(int resId) {
    return getContext().getString(resId);
  }
}

File: app/src/main/java/com/teixeira0x/subtypo/ui/prefs/UIModePreference.kt
package com.teixeira0x.subtypo.ui.prefs

import android.content.Context
import android.util.AttributeSet
import androidx.preference.Preference
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.teixeira0x.subtypo.App
import com.teixeira0x.subtypo.R

class UIModePreference : Preference {

  constructor(context: Context) : super(context)

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

  constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int,
  ) : super(context, attrs, defStyleAttr)

  constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int,
    defStyleRes: Int,
  ) : super(context, attrs, defStyleAttr, defStyleRes)

  override fun onClick() {
    super.onClick()

    val items =
      arrayOf(
        getString(R.string.pref_appearance_ui_mode_value_follow),
        getString(R.string.pref_appearance_ui_mode_value_light),
        getString(R.string.pref_appearance_ui_mode_value_dark),
      )

    val prefs = sharedPreferences ?: return
    val editor = prefs.edit()
    val selectedItemIndex = prefs.getInt(key, 0)

    MaterialAlertDialogBuilder(context)
      .setTitle(title)
      .setSingleChoiceItems(items, selectedItemIndex) { _, w ->
        editor.putInt(key, w)
      }
      .setNegativeButton(R.string.cancel, null)
      .setPositiveButton(R.string.save) { _, _ ->
        editor.apply()
        App.instance.updateUIMode()
      }
      .show()
  }

  private fun getString(resId: Int): String {
    return context.getString(resId)
  }
}
